hilti/toolchain/src/compiler/visitors/normalizer.cc

```diff
@@ -0,0 +1,209 @@
+                                                      {std::move(map), std::move(key), std::move(value)},
+                                                      assign.meta());
+
+            logChange(p.node, index_assign);
+            p.node = std::move(index_assign);
+            modified = true;
+            return;
+        }
+
+        // Rewrite assignments involving struct elements to use the non-const member operator.
+        if ( auto member_const = lhs.tryAs<operator_::struct_::MemberConst>() ) {
+            auto new_lhs = operator_::struct_::MemberNonConst::Operator().instantiate(member_const->operands().copy(),
+                                                                                      member_const->meta());
+            Expression n = expression::Assign(new_lhs, assign.source(), assign.meta());
+            logChange(p.node, n);
+            p.node = std::move(n);
+            modified = true;
+            return;
+        }
+
+        // Rewrite assignments involving tuple ctors on the LHS to use the
+        // tuple's custom by-element assign operator. We need this to get
+        // constness right.
+        auto lhs_ctor = lhs.tryAs<expression::Ctor>();
+        if ( lhs_ctor && lhs_ctor->ctor().isA<ctor::Tuple>() ) {
+            if ( expression::isResolved(assign.source()) && expression::isResolved(assign.target()) ) {
+                auto n = operator_::tuple::CustomAssign::Operator().instantiate({assign.target(), assign.source()},
+                                                                                assign.meta());
+                logChange(p.node, n);
+                p.node = std::move(n);
+                modified = true;
+                return;
+            }
+        }
+    }
+
+    void operator()(const statement::If& n, position_t p) {
+        if ( n.init() && ! n.condition() ) {
+            auto cond = expression::UnresolvedID(n.init()->id());
+            logChange(p.node, cond);
+            p.node.as<statement::If>().setCondition(std::move(cond));
+            modified = true;
+        }
+    }
+
+    void operator()(const statement::Switch& s, position_t p) { p.node.as<statement::Switch>().preprocessCases(); }
+
+    void operator()(const type::Library& t, position_t p) {
+        auto& type = p.node.as<Type>();
+
+        if ( ! type.cxxID() )
+            // Make it equal to types with the same C++ representation.
+            type.setCxxID(ID(t.cxxName()));
+    }
+
+    void operator()(const type::Struct& t, position_t p) { type::Struct::setSelf(&p.node); }
+};
+
+} // anonymous namespace
+
+struct VisitorClearCanonicalIDs : public visitor::PreOrder<void, VisitorClearCanonicalIDs> {
+    result_t operator()(const Declaration& d, position_t p) { p.node.as<Declaration>().setCanonicalID(ID()); };
+};
+
+struct VisitorComputeCanonicalIDs : public visitor::PreOrder<ID, VisitorComputeCanonicalIDs> {
```

Comment:
Like discussed offline, this seems to compute different IDs for method definitions and their declarations, e.g., given a module

```
module bar;
public type X = unit {
    on %done {}
};
```

I see a method declaration with `[canon-id: bar::bar::X::__on_0x25_done]`

```
[debug/ast-final]       | bar::X::__on_0x25_done -> declaration::Function %241 <linkage="struct" parent_type="%1"> [canon-id: bar::bar::X::__on_0x25_done] [@d:7fb0f3c35520] ([@d:7fb0f3c35520])
```

for the field declaration `[canon-id: bar::X::__on_0x25_done]`

```

```
declaration::Field <cc="<standard>"> (bla.spicy:2:17-4:2) [canon-id: bar::X::__on_0x25_done] [@d:7fb0f3c35e70]
```

Even though these are different declarations, semantically they refer to the same entity.

Not necessarily an issue, but related it also seems hard to connect these to the related `UnitHook` which has neither a canon-id nor any other fields which seems to tie it to the field or function declaration (at least in what is printed to the ast debug stream).

Reply:
This has been fixed.